### PR TITLE
FDB Builtin reduce

### PR DESCRIPTION
**Pull request by garrensmith**
## Overview

Adds builtin reduce indexes for CouchDB on FDB. The reduce indexes are created using a skiplist like data structure. See the [RFC](https://github.com/apache/couchdb-documentation/pull/441) for the full details. 

This work is not 100% complete. Currently, the reduce index cannot update an index when a document is updated as it will not remove the original document from the index. 

There are two ways to query the index. By default a query of the reduce index will read through the full skiplist to return the results. But if you add the parameter `use_skiplist=false` it will only read from the level = 0 of the skip list and compute the group_level results in memory. level = 0 of the skiplist is equivalent of `group=true` for reduce.

I'm currently at the point where I need feedback from the rest of the team on how best to finish the work. The Skiplist algorithm works but it does seem a bit slow and its pretty complex. I need help on the following things:

1. Review the skip list traversing algorithm to see if we can improve it and make it faster
1. Review the add keys to the skip list and see if there are any bugs or issues with this
1. Finally should we continue with the skip list implementation? Or should we rather go with a reduce index that just stores the equivalent of the level 0 keys for the skiplist, and then calculate the reduce group levels in memory when returning the results. 

## Testing recommendations

New tests have been added. Current the PR fails the reduce_builtin.js when using the full skiplist but pass with only using level 0.

## Checklist

- [ ] Code is written and works correctly
- [ ] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] A PR for documentation changes has been made in https://github.com/apache/couchdb-documentation

### CODE REVIEW

File: src/couch_views/src/couch_views_reduce.erl [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+            lists:map(fun({_, ReduceFun}) ->
+                couch_views_reducer:reduce(ReduceFun, Result)
+            end, ViewReduceFuns)
+
+        end, lists:zip(Views, Results)),
+
+        MappedResult#{
+            reduce_results => ReduceResults
+        }
+    end, MappedResults).
```
[the remark this answers is not in the document]
Comment (garrensmith) [reply src/couch_views/src/couch_views_reduce.erl]: I would prefer to leave it more verbose. Its something we have tried to do more in the couch_views map work. I find it a little easier to read.

**Comment by garrensmith** [comment]
@chewbranca thanks for the review. I've made the fixes you suggested. Where I would really value your feedback is the actual skip list algorithm. `couch_views_skiplist.erl`. I need another set of eyes to check if it looks correct and if you can see any improvements.

**Comment by chewbranca** [comment]
So I've been perusing the RFC, `couch_views_skiplist.erl` and the original whitepaper, and one major difference jumps out to me: the skip List implementation here is _not_ a linked list based approach, as per traditional skip Lists, but rather embeds the levels into the keys themselves to allow for walking different levels of values. The other thing that jumps out to me is that the querying algorithm utilizes a bottom up approach starting at level zero rather than a top down approach starting at max level, which seems backwards and potentially less performant, especially since right now it seems to spend a lot of time fetching individual keys.

I think storing the levels in the key affords us an opportunity to heavily lean on range queries, and potentially ignore direct key lookups entirely. Essentially I was thinking you could start at max level, take the baseline supplied startkey and endkey, and then do a range query on `{...,...,$MAX_LEVEL,$start_key}..{...,...,$MAX_LEVEL,$end_key}`. Then you take the results and divide and conquer the lower and upper key spaces with similar queries at one level lower, eg lower query of `{...,...,$MAX_LEVEL - 1,$start_key}..{...,...,$MAX_LEVEL - 1,$results[0]}` and upper query of `{...,...,$MAX_LEVEL - 1,$results[-1]}..{...,...,$MAX_LEVEL - 1,$end_key}`. And then you recurse through the lower levels until you've bottomed out or `$start_key == $results[0] and $end_key == $results[-1]`. In theory this algorithm could fulfill the requests without ever needing to do individual key lookups, while maximizing the range of results queried at the upper levels.

Here's a proof of concept query algorithm:

```erlang
query(View, StartKey, EndKey) ->
    query(View, ?MAX_LEVEL, StartKey, EndKey).


query(_, ?MIN_LEVEL, _, _) ->
    [];
query(View, Level, StartKey, EndKey) ->
    LevelDown = Level - 1,
    Start = make_key(View, Level, StartKey),
    End = make_key(View, Level, EndKey),

    [First | _] = Results = fdb_range_query(Start, End),
    Last = lists:last(Results),

    case {StartKey < First, Last < EndKey} of
        {true, true} ->
            query(View, LevelDown, StartKey, First) ++ Results ++ query(View, LevelDown, Last, EndKey);
        {true, false} when Last =:= EndKey ->
            query(View, LevelDown, StartKey, First) ++ Results;
        {false, true} when First =:= StartKey ->
            Results ++ query(View, LevelDown, Last, EndKey);
        {false, false} when First =:= StartKey andalso Last =:= EndKey ->
            Results;
        {_, _} ->
            throw({error, shouldnt_be_here})
    end.


make_key(View, Level, Key) ->
    {View#view.db, ?DB_VIEWS, View#view.sig, ?VIEW_REDUCE_SK_RANGE, View#view.id, Level, Key}.
```

It's quite possible this approach isn't practical for any number of reasons, but it jumped out to me as a potentially viable algorithm that is simple and utilizes the range query support provided by FoundationDB>

**Comment by chewbranca** [comment]
That approach would also easily support the non skip list toggle with something like

```erlang
query_view(View, StartKey, EndKey, UseSkipList) ->
    case UseSkipList of
        true -> query(View, StartKey, EndKey);
        false -> query(View, ?MIN_LEVEL, StartKey, EndKey)
    end.
```

**Comment by garrensmith** [comment]
@chewbranca I'm not sure this would work. I think the algorithm treats the skiplist a bit too much like a b-tree. I used your algorithm and did a bit of a mental run-through of the skiplist diagram from the [rfc](https://github.com/apache/couchdb-documentation/blob/8864c35dcb27c9e5996532255fa59e544dee885d/rfcs/012-fdb-reduce.md#create)

There were two things that were issues that I couldn't quite see a way around. The first one is that once we reach an endkey at a level above 0 it assumes it can stop. The issue here is that any node above level 0 contains aggregations of the keys a level below. So we can easily have aggregations of key/values that are not in the group level for that key or not in the range for the query. So far I've found we always have to go down to level 0 to check that an endkey value or group level key has the correct values. So basically a skip list allows us to skip many keys for a group level but we still have to check the final key in the group level to make sure the value does not contain aggregates of values outside the group level. 

The second issue is that we would have to keep all the results in memory before doing a final rereduce right at the end. Maybe that isn't an issue, but I think on large reduce queries we will OOM on this. We could rereduce after each level read and that should reduce the number of results. But it will again require some logic to make sure all the values we have are correct.

The bottom-up design I have is definitely not the most optimal, so its great to have you thinking of different designs. But with the bottom-up design, we do find the end keys correctly and can stream group level keys to a user as we traverse the skiplist.

**Comment by davisp** [comment]
I finally had some downtime to start looking into this more closely. So far I've only spent a couple hours reading through and thinking about things so hopefully this comes out semi-coherently.

So far I'm just focusing on the skiplist implementation. First off, as @chewbranca's confusion points out, we should probably scrub all references to skiplists because these are different enough that it'll just lead to confusion when people attempt to compare them. I'd at least mention RankedSets and include links to the Java implementation in fdb-record-layer [1]. RankedSets is a bit overly specific given we've got arbitrary reduce functions. But maybe `reduced_sets` would be good enough?

I started reading on the update side of things first and it looks like you've basically nailed it. I think we should change around a few definitions though. If we refer to the map index's key/value pairs as level 0 and then keep the rest as written we'll have covered custom javascript reduce functions and all of the builtin reduce functions. Which is way more awesome than when I was thinking we were going to have to invent some sort of un-reduce function which would have been super error prone. Brilliant work there!

The query/traversal side of things is going to need some work. I think there's a bit of confusion around how group levels work that got mixed into the traversal code. To simplify the implementation and development here you should write a function that will correctly answer a group=true query with a start and end key. Once you have an implementation of that function, turning it into a `foldl` style API will be straightforward. And then once you have that API you'll basically just be copying the group logic from couch_btree. The algorithm that @chewbranca sketched should be a good starting place here.

Once you've got all that done, adding the removal logic is as simple as cribbing off the Java `RankedSet.remove` method.

That's about it from me for now. Good work!

[1] https://github.com/FoundationDB/fdb-record-layer/blob/master/fdb-extensions/src/main/java/com/apple/foundationdb/async/RankedSet.java

**Comment by davisp** [comment]
Also I should have mentioned, that algorithmically the example that @chewbranca is a good starting place. @garrensmith was quite right that we'll have to worry about RAM usage and what not. Eventually, the `fdb_range_query` part will be limited in the number of rows returned and then we'll have to handle that complication. However, for the first pass I'd ignore those concerns and write it as simply as currently sketched.

**Pull request closed without merging**